content/blog/angular-extend-class/index.md

```diff
@@ -0,0 +1,842 @@
+
+```
+TS2415: Class 'AppComponent' incorrectly extends base class 'BaseComponent'.
+   Types have separate declarations of a private property 'document'.
+```
+
+To solve this, we simply need to make our `BaseComponent`'s `constructor` properties `public` instead of `private`:
+
+```typescript
+@Component({
+  template: ''
+})
+abstract class BaseComponent implements OnInit {
+  constructor(@Inject(DOCUMENT) public document: Document) {}
+  ngOnInit() {
+    console.log(document.title);
+  }
+}
+
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>Test</p>
+  `,
+})
+class AppComponent extends BaseComponent implements OnInit {
+  constructor(@Inject(DOCUMENT) public override document: Document) {
+    super(document);
+    console.log(document.body);
+  }
+}
+```
+
+> Remember to keep your `override` property in the `AppComponent` `constructor`, otherwise you'll have errors.
+
+# Why you don't want to extend Angular base classes {#dont-extend-base-classes}
+
+Now that we've learned how to extend base classes in Angular to share lifecycle methods, allow me to flip the script:
+
+**You shouldn't use a base class in Angular**.
+
+<video title="A shocked sock puppet monkey" src="./shocked-monkey-gif.mp4"></video>
+
+Why?
+
+Well, it's often sited by Angular experts that using a base class is brittle and difficult to maintain.
+
+For example, let's say that you have a base component that doesn't use dependency injection, but then suddenly need to add dependency injection. What do you do?
+
+Well, you'd have to refactor every instance that you extended that class.
+
+Similarly, if you add a lifecycle method that you want to overwrite in the future, there can be someheadaches depending in which order you do things in.
+
+There are more than a few ways to write this code in a different, more stable, way.
+
+## Fixing things the right way {#the-fix}
+
+There is a better way to write this code differently today that solve the problems of maintainability a bit better.
+
+Namely, using an `@Injectable` class that's provided on a per-class level enables you to have explicit `setup` and `takedown` functions that you call manually:
+
+```typescript
+@Injectable()
+class WindowSizeService {
+  height = window.innerHeight;
+  width = window.innerWidth;
+ 
+  onResize = () => {
+    this.height = window.innerHeight;
+    this.width = window.innerWidth;
+  }
+
+  addListeners() {
+    window.addEventListener('resize', this.onResize);
+  }
+
+  removeListeners() {
+    window.removeEventListener('resize', this.onResize);
+  }
+}
+
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>The window is {{windowSize.height}}px high and {{windowSize.width}}px wide</p>
+  `,
+  providers: [WindowSizeService]
+})
+class AppComponent implements OnInit, OnDestroy {
+  constructor(public windowSize: WindowSizeService) {
+  }
+
+  ngOnInit() {
+    this.windowSize.addListeners();
+  }
+
+  ngOnDestroy() {
+    this.windowSize.removeListeners();
+  }
+}
+```
```

Comment:
Calling any kind of hook manually, that is a cardinal sin or at the very least a code smell in Angular apps 😬

You are correct in using a component-level provided service but use the `OnDestroy` hook and the service constructor as the `OnInit`:
```typescript
@Injectable()
class WindowSizeService implements OnDestroy {
  height = this.window.innerHeight;
  width = this.window.innerWidth;
 
  constructor(@Inject(DOCUMENT) private document: Document) {
    this.window = document.defaultView!;
    this.window.addEventListener('resize', this.onResize);
  }
  
  ngOnDestroy() {
    this.window.removeEventListener('resize', this.onResize);
  }

  onResize = () => {
    this.height = this.window.innerHeight;
    this.width = this.window.innerWidth;
  }
}
```

Alternatively, use an [`APP_INITIALIZER`](https://angular.io/api/core/APP_INITIALIZER) or an [`APP_BOOTSTRAP_LISTENER`](https://angular.io/api/core/APP_BOOTSTRAP_LISTENER) instead of the service constructor.